File: cpp/src/benchmarks/dictionnary.cpp
#include "dictionnary.h"
#include "tools/error.h"
#include <fstream>
#include <iostream>
#include <map>
#include <string>
#include <string_view>

namespace bch {

void Node::record(std::string_view::const_iterator first,
                  std::string_view::const_iterator last)
{
    if (first == last) {
        is_terminal = true;
        return;
    }
    Node &next = children[*first];
    next.record(first + 1, last);
}
void Dict::record(std::string_view str)
{
    root.record(str.begin(), str.end());
}

void Dict::print()
{
    std::cout << "Dictionnary content:\n";
    print_node(root, "");
    std::cout.flush();
}

void Dict::print_node(Node const &node, std::string str)
{
    if (node.is_terminal) {
        std::cout << "  --> " << str << '\n';
    }
    for (const auto &[ch, next] : node.children) {
        print_node(next, str + ch);
    }
}

void benchmark_dictionnary()
{
    Dict          dict;
    std::ifstream in(DATA_DIR "/dict.txt");
    if (!in) {
        throw IOError("File \"dict.txt\" not found!");
    }
    std::string buffer;
    while (!in.eof()) {
        in >> buffer;
        dict.record(buffer);
    }
}

} // namespace bch

File: rust/src/benchmarks/dictionnary.rs
use std::collections::BTreeMap;
use std::str::Chars;
use super::super::tools::paths::get_data_path;

#[derive(Default, Clone)]
pub struct Node {
    pub is_terminal: bool,
    children: BTreeMap<char, Node>
}

#[derive(Default, Clone)]
pub struct Dict {
    pub root: Node
}

impl Node {
    pub fn record(&mut self, word: &mut Chars)
    {
        let c: char;
        match word.next() {
            Some(ch) => c = ch,
            None => {
                self.is_terminal = true;
                return;
            }
        };
        let node = self.children.entry(c).or_insert(Node::default());
        node.record(word);
    }

    pub fn contains(&self, word: &mut Chars) -> bool
    {
        match word.next() {
            Some(ch) => {
                let next = self.children.get(&ch);
                if next.is_some() {
                    return next.unwrap().contains(word);
                }
                else {
                    return false;
                }
            },
            None => return self.is_terminal
        }
    }
}

impl Dict {
    pub fn record(&mut self, str: &str)
    {
        self.root.record(&mut str.chars());
    }

    pub fn contains(&self, str: &str) -> bool
    {
        self.root.contains(&mut str.chars())
    }

    // fn print_node(node: &Node, str: String)
    // {
    //     if node.is_terminal {
    //         println!(" --> {}", &str);
    //     }
    //     for (ch, next) in &node.children {
    //         let mut str_cpy = str.clone();
    //         str_cpy.push(*ch);
    //         Dict::print_node(&next, str_cpy);
    //     }
    // }

    // pub fn print(&self)
    // {
    //     println!("Dictionnary content:");
    //     Dict::print_node(&self.root, "".to_string());
    // }
}

use std::io::{BufRead, BufReader};
use std::fs::File;

pub fn benchmark_dictionnary() -> ()
{
    let mut dict = Dict::default();
    let mut path = get_data_path();
    path.push("dict.txt");
    let reader = BufReader::new(File::open(&path).expect("Cannot open dict.txt"));

    for line in reader.lines() {
        dict.record(&line.unwrap());
    }
    assert!(dict.contains("black-hole"));
    assert!(!dict.contains("white-hole"));
}
